src/osd/modules/netdev/vmnet.cpp

```diff
@@ -0,0 +1,288 @@
+// license:BSD-3-Clause
+// copyright-holders:Kelvin Sherlock
+/*
+    macOS vmnet network interface, direct version
+    by Kelvin Sherlock
+
+    vmnet provides an easy way for emulators and virtual machines to give their guest systems
+    access to the outside LAN/Internet.  But for good security reasons it requies you to have
+    root privlidges.
+
+    This module does not attempt to solve that.  MAME must be run as root (or sudo) in order
+    to have network access, and all of the issues that implies for security and ease of use
+    are in play.
+
+    ---
+
+    https://developer.apple.com/documentation/vmnet
+
+    A sandboxed user space process must have the com.apple.vm.networking entitlement
+    in order to use vmnet API.
+*/
+
+#if defined(OSD_NET_USE_VMNET)
+
+#include "emu.h"
+#include "osdcore.h"
+#include "osdnet.h"
+#include "modules/osdmodule.h"
+#include "netdev_module.h"
+
+#include <cstdint>
+#include <cstdlib>
+#include <cstdio>
+
+#include <sys/types.h>
+#include <sys/uio.h>
+#include <vmnet/vmnet.h>
+
+#include "vmnet_common.h"
+
+namespace osd {
+
+namespace {
+
+int classify_mac(uint8_t *mac) {
+	if ((mac[0] & 0x01) == 0) return 1; /* unicast */
+	if (memcmp(mac, "\xff\xff\xff\xff\xff\xff", 6) == 0) return 0xff; /* broadcast */
+	return 2; /* multicast */
+}
+
+class vmnet_module : public osd_module, public netdev_module
+{
+public:
+
+	vmnet_module() : osd_module(OSD_NETDEV_PROVIDER, "vmnet"), netdev_module() {}
+	virtual ~vmnet_module() {}
+
+	virtual int init(osd_interface &osd, const osd_options &options);
+	virtual void exit();
+
+	virtual bool probe() {
+		return true;
+	}
```

Comment:
This module won’t work without appropriate entitlements.  Can you make this check for the `com.apple.vm.networking` entitlement or running as root, and return `false` if neither is the case?

You can check entitlements with something like:
```c++
	SecCodeRef me;
	if (SecCodeCopySelf(kSecCSDefaultFlags, &me) == errSecSuccess)
	{
		SecStaticCodeRef static_code;
		if (SecCodeCopyStaticCode(me, kSecCSDefaultFlags, &static_code) == errSecSuccess)
		{
			CFDictionaryRef signing;
			if (SecCodeCopySigningInformation(static_code, kSecCSDefaultFlags, &signing) == errSecSuccess)
			{
				CFDictionaryRef entitlements = CFDictionaryRef(CFDictionaryGetValue(signing, kSecCodeInfoEntitlementsDict));
				if (entitlements)
				{
					// value should be 1 if entitlement is available, but the key shouldn't be present at all otherwise
					if (CFDictionaryGetValue(entitlements, CFSTR("com.apple.vm.networking")))
					{
						// entitlement should be available if you get here.
					}
				}
				CFRelease(signing);
			}
			CFRelease(static_code);
		}
		CFRelease(me);
	}
```

You should be able to check if you’re running as root by getting the current uid/euid the usual UNIX way.

Reply:
Done.